lib/Parse/ParsePattern.cpp

```diff
@@ -242,26 +238,47 @@ Parser::parseParameterClause(SourceLoc &leftParenLoc,
       }
 
       // (':' ('inout')? type)?
+      // also handle the case of (':' ('let' or 'var')? type)? as errors
+      // and give diagnostics for that case
       if (consumeIf(tok::colon)) {
 
         SourceLoc postColonLoc = Tok.getLoc();
 
         bool hasDeprecatedInOut =
           param.SpecifierKind == ParsedParameter::InOut;
-
-        if (Tok.is(tok::kw_inout)) {
-          param.LetVarInOutLoc = consumeToken();
-          param.SpecifierKind = ParsedParameter::InOut;
-          if (hasDeprecatedInOut) {
-            diagnose(param.LetVarInOutLoc, diag::inout_as_attr_deprecated)
-              .fixItRemove(param.LetVarInOutLoc);
+        bool hasValidInOut = false;
+        
+        while (Tok.isAny(tok::kw_inout, tok::kw_let, tok::kw_var)) {
+          if (Tok.is(tok::kw_var)) {
+            diagnose(Tok.getLoc(), diag::parameter_var_as_attr)
+            .fixItRemove(Tok.getLoc());
+            consumeToken();
+            param.isInvalid = true;
+          } else if (Tok.is(tok::kw_let)) {
+            diagnose(Tok.getLoc(), diag::parameter_let_as_attr).
+            fixItRemove(Tok.getLoc());
```

Comment:
I don't think that you need to catch the case of let/var after the colon, particularly since these are going to be removed from swift parameters in general.  However, if you really want to handle this, please merging the two parameter_var_as_attr/parameter_let_as_attr diagnostics into a single one that takes a bool (to tell it whether to print let or var) which it can dispatch on with %select(let|var)0 in the diagnostic text.